Add towncrier to maintain release notes (#1303)
cocotb/cocotb: documentation/source/newsfragments/README.rst

:orphan:

Release Notes
=============

We are using `towncrier <https://pypi.org/project/towncrier/>`_ to handle
our release notes, and this directory contains the input for it -
"news fragments" which are short files that contain a small
**ReST**-formatted text that will be added to the next version's
Release Notes page.

Each file should be named like ``<ISSUE_OR_PR>.<TYPE>.rst``,
where ``<ISSUE_OR_PR>`` is an issue or a pull request number -
whatever is most useful to link to,
and ``<TYPE>`` is one of:

* ``feature``: New user-facing feature.
* ``bugfix``: A bug fix.
* ``doc``: Documentation improvement.

> * ``removal``: Deprecation or removal of public API or behavior.

Reviewer: I think we might want a category for behavior changes, such as #1375 which isn't really a feature, but is a user-visible change that might be worth at least mentioning

Author: I just added that locally and will push once my network problems are over.

Author: ``change`` is now in.

**ReST**-formatted text that will be added to the next version's
Release Notes page.

Each file should be named like ``<ISSUE_OR_PR>.<TYPE>.rst``,
where ``<ISSUE_OR_PR>`` is an issue or a pull request number -
whatever is most useful to link to,
and ``<TYPE>`` is one of:

* ``feature``: New user-facing feature.
* ``bugfix``: A bug fix.
* ``doc``: Documentation improvement.
* ``removal``: Deprecation or removal of public API or behavior.
* ``change``: A change in public API or behavior.

In that file, make sure to use full sentences with correct case and punctuation,
and do not use a bullet point at the beginning of the file.
Use Sphinx references (see https://sphinx-tutorial.readthedocs.io/cheatsheet/)
if you refer to added classes, methods etc.

An example file could consist of the content between the marks:

>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
Summary of a new feature.

This is a second paragraph.
<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

Note that the last paragraph should be a normal sentence and not e.g. code,
because the issue number is appended there.

> Towncrier runs automatically when you run any of the Sphinx output build commands.

Reviewer: I thought I commented on this but maybe I forgot to submit the comment.

```suggestion
Towncrier automatically assembles a list of unreleased changes when building sphinx.

When performing a release, it should be run independently, which will delete all the merged newsfragments, and create new commits.
```
or something.

Author: Added, thanks.

:orphan:

Release Notes
=============

We are using `towncrier <https://pypi.org/project/towncrier/>`_ to handle

> our relase notes, and this directory  contains the input for it -

Reviewer: ```suggestion
our release notes, and this directory contains the input for it -
```

"news fragments" which are short files that contain a small
**ReST**-formatted text that will be added to the next version's
Release Notes page.

Make sure to use full sentences with correct case and punctuation,
and use Sphinx references

> (see https://sphinx-tutorial.readthedocs.io/cheatsheet/).

Reviewer: Do we want:
```
TheFeature
----------
Summary of new feature
```
or

```
* Summary of new feature
```
or 
```
Summary of new feature
```

as the format in each fragment? We should give an example of whatever we pick in this file (I don't care which)

Author: Added (the last one since the bullet point is added by towncrier in our configuration).

* ``feature``: New user-facing feature.
* ``bugfix``: A bug fix.
* ``doc``: Documentation improvement.
* ``removal``: Deprecation or removal of public API.
* ``misc``: A ticket has been closed, but it is not of interest to users.

.. comment_numpy_has_these:
    * ``new_function``: New user facing functions.
    * ``deprecation``: Changes existing code to emit a DeprecationWarning.
    * ``future``: Changes existing code to emit a FutureWarning.
    * ``expired``: Removal of a deprecated part of the API.
    * ``compatibility``: A change which requires users to change code and is not
      backwards compatible. (Not to be used for removal of deprecated features.)
    * ``c_api``: Changes in the Numpy C-API exported functions
    * ``new_feature``: New user facing features like ``kwargs``.
    * ``improvement``: Performance and edge-case changes
    * ``change``: Other changes
    * ``highlight``: Adds a highlight bullet point to use as a possibly highlight
      of the release.

You can install ``towncrier`` and run ``towncrier --draft``
if you want to get a preview of how your change will look in the final release

> notes.

Reviewer: I think we need to add a brief section somewhere as instructions for maintainers about how to assemble a release.

Author: Once this PR is accepted I will add that information to https://github.com/cocotb/cocotb/wiki/Release-Process

Reviewer 2: Why doesn't towncrier always run when creating the docs? This way we can also see if the fragments were added correctly by previewing the RTD page that's generated for each PR.

Author: I'm adding that.

> :orphan:

Reviewer 2: We tried to move documentation to one sentence per line style. How about doing that for this file as well?

Author: I would not follow that rule strictly in cases where it still leads to longish lines, but punctuation is in general a good candidate for line breaks.
In the case of this specific file, the ReST is not rendered but read in raw form, so should be pleasant to read in raw as well.

Release Notes
=============

We are using `towncrier <https://pypi.org/project/towncrier/>`_ to handle
our release notes, and this directory contains the input for it -
"news fragments" which are short files that contain a small
**ReST**-formatted text that will be added to the next version's
Release Notes page.

Each file should be named like ``<PULL REQUEST>.<TYPE>.rst``, where

> ``<PULL REQUEST>`` is a pull request number, and ``<TYPE>`` is one of:

Reviewer 2: In many cases, the issue is the start and end of a discussion, and the PR only implements it. Of course, that's not always true, and not every PR has an associated issue.

In case where the issue is the main center of a discussion (as it should be), then the issue number is the one to use here. Can you reword that somehow to allow both issue and PR numbers?

Reviewer: What matters most is that the number is unique, but agreed that sometimes the issue is more valuable.

It's usually trivial to get from the PR to the issue though through github anyway, so I'd be inclined to leave this advice as is.

* ``feature``: New user-facing feature.
* ``bugfix``: A bug fix.
* ``doc``: Documentation improvement.

> * ``removal``: Deprecation or removal of public API.

Reviewer 2: API -> behavior

(API is a common term for libraries, but cocotb has more than that, e.g. the Makefiles)

Reviewer: ```suggestion
* ``removal``: Deprecation or removal of public API or behavior.
```

> * ``misc``: A ticket has been closed, but it is not of interest to users.

Reviewer 2: If something is not of interest to a user, then it doesn't need to appear in the release notes.

.. comment_numpy_has_these:
    * ``new_function``: New user facing functions.
    * ``deprecation``: Changes existing code to emit a DeprecationWarning.
    * ``future``: Changes existing code to emit a FutureWarning.
    * ``expired``: Removal of a deprecated part of the API.
    * ``compatibility``: A change which requires users to change code and is not
      backwards compatible. (Not to be used for removal of deprecated features.)
    * ``c_api``: Changes in the Numpy C-API exported functions
    * ``new_feature``: New user facing features like ``kwargs``.
    * ``improvement``: Performance and edge-case changes
    * ``change``: Other changes
    * ``highlight``: Adds a highlight bullet point to use as a possibly highlight

>       of the release.

Reviewer 2: What's the purpose of this section? Can it be removed?

Reviewer: It's just a comment, which is there for people who come back to this and think about adding new categories and need inspiration. We could replace it with a link to the numpy one to make this shorter.

Author: Right, it's meant for this discussion mainly - are we okay with the categories that towncrier supports by default, or do we want different ones? For example, the deprecation vs. removal differentiation.

In that file, make sure to use full sentences with correct case and punctuation,
and do not use a bullet point at the beginning of the file.
Use Sphinx references
(see https://sphinx-tutorial.readthedocs.io/cheatsheet/) if you refer
to added classes, methods etc.

> An example line could be ``Summary of new feature.``.

Reviewer: Are you describing the first line of the file, or its entire contents? Are multiple paragraphs supported?

Reviewer 2: Can you give a full example of such a file?

You can install ``towncrier`` and run ``towncrier --draft``
if you want to get a preview of how your change will look in the final release
notes.

.. note::

    This README was adapted from the numpy changelog readme which in turn

>     took it from pytest under the terms of the MIT licence.

Reviewer 2: You can remove that paragraph, the text is trivial enough.
